### ISSUE TITLE
Remove mktap and corresponding twistd options (mktap is broken in trunk)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5293]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)| @teratorn reported|
|-|-|
|Trac ID|trac#5293|
|Type|defect|
|Created|2011-10-06 04:54:10Z|
|Branch|https://github.com/twisted/twisted/tree/remove-mktap-5293-2|

Using a fresh copy of SVN trunk:

```
% mktap web
/home/teratorn/code/Twisted/trunk/bin/mktap:11: DeprecationWarning: mktap and related support modules are deprecated as of Twisted 8.0.  Use Twisted Application Plugins with the 'twistd' command directly, as described in 'Writing a Twisted Application Plugin for twistd' chapter of the Developer Guide.
  from twisted.scripts.mktap import run
Traceback (most recent call last):
  File "/home/teratorn/code/Twisted/trunk/bin/mktap", line 12, in <module>
    run()
  File "/home/teratorn/code/Twisted/trunk/twisted/scripts/mktap.py", line 182, in run
    options['uid'], options['gid'])
  File "/home/teratorn/code/Twisted/trunk/twisted/scripts/mktap.py", line 77, in addToApplication
    sob.IPersistable(a).save(filename=append, passphrase=passphrase)
  File "/home/teratorn/code/Twisted/trunk/twisted/persisted/sob.py", line 122, in save
    self._saveTemp(filename, passphrase, dumpFunc)
  File "/home/teratorn/code/Twisted/trunk/twisted/persisted/sob.py", line 93, in _saveTemp
    dumpFunc(self.original, f)
  File "/home/teratorn/code/Twisted/trunk/twisted/persisted/sob.py", line 106, in dumpFunc
    pickle.dump(obj, file, 2)
cPickle.UnpickleableError: Cannot pickle <type 'select.poll'> objects
```

No problem here with the last release, 11.0.

Setting to 11.1 milestone since shipping totally broken commands (assuming it's not just me) would be really bad form.

<details><summary>Searchable metadata</summary>

```
trac-id__5293 5293
type__defect defect
reporter__teratorn teratorn
priority__normal normal
milestone__Twisted_11_1 Twisted-11.1
branch__branches_remove_mktap_5293_2 branches/remove-mktap-5293-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1317876850000000 1317876850000000
changetime__1319202952000000 1319202952000000
version__None None
owner__exarkun exarkun
cc__itamarst cc__jesstess cc__teratorn@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

While we don't have deprecation warnings, the documentation has suggested not using mktap for a long time now (which seems like a better deprecation method than warnings for programs, as opposed to APIs). Maybe we should just delete it and the corresponding twistd options.

The actual change that broke it is probably making poll the default reactor on Linux. Unless that is interacting with some other change, that was merged 5 months ago -- the fact no one noticed until now is telling.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

We do have deprecation warnings, in fact.  They have been in place since Twisted 8.  I think we should remove mktap.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

OK then. I updated the summary.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

Oh, and, we should also remove the references to it in the documentation.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [32885]) Branching to 'remove-mktap-5293'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

I suggest just removing all references to mktap, period. We've almost certainly expunged all references to TAP files by now, and assuming our user base is still growing most users have never even used mktap. So no point mentioning it; it doesn't clear up any confusion, it just adds confusion.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(wrote this comment before I saw itamar's comment)

I removed it and most references to it.  I left a couple, because they're of historical inherest or they refer to functionality we forgot to deprecate when we deprecated mktap (see [#5317](https://github.com/twisted/twisted/issues/5317)).

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/remove-mktap-5293)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set owner to @exarkun|
|-|-|

Thanks for working on this, exarkun. There are some references to mktap in the ZSH completion code that should probably also go away:

```
./twisted/python/_twisted_zsh_stub:#compdef trial conch mktap cftp tapconvert twistd ckeygen lore pyhtmlizer tap2deb tkconch manhole tap2rpm
./twisted/python/_twisted_zsh_stub:        autoload +X _trial _conch _mktap _cftp _tapconvert _twistd _ckeygen
./twisted/python/zsh/_mktap:#compdef mktap
./twisted/python/zshcomp.py:    Builder for the mktap command
./twisted/python/zshcomp.py:               ('mktap', 'twisted.scripts.mktap', 'FirstPassOptions'),
./twisted/python/zshcomp.py:specialBuilders = {'mktap'  : MktapBuilder,
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Mmph, yea, those.  I expect teratorn is going to get rid of _most_ of those references when he merges [#3078](https://github.com/twisted/twisted/issues/3078).  However, maybe the `_twisted_zsh_stub` does need some special attention here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Oops.  But `_twisted_zsh_stub` got renamed in the [[#3078](https://github.com/twisted/twisted/issues/3078)](https://github.com/twisted/twisted/issues/3078) branch.  Maybe I should just ask teratorn to fix account for mktap's deletion there.  On the other hand, the #3078 branch still has completion files for `websetroot` and `_tkmktap`, but those were deleted a long time ago too.  I'm not really sure what's going on with those.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_12"></a>|@teratorn commented|
|-|-|

Replying to [exarkun](#note_11):
> Oops.  But `_twisted_zsh_stub` got renamed in the [[#3078](https://github.com/twisted/twisted/issues/3078)](https://github.com/twisted/twisted/issues/3078) branch.  Maybe I should just ask teratorn to fix account for mktap's deletion there.  On the other hand, the #3078 branch still has completion files for `websetroot` and `_tkmktap`, but those were deleted a long time ago too.  I'm not really sure what's going on with those.

The "stub" completion file, _twisted_zsh_stub (renamed to twisted-completion.zsh in [[[#3078](https://github.com/twisted/twisted/issues/3078)](https://github.com/twisted/twisted/issues/3078)](https://github.com/twisted/twisted/issues/3078)) no longer contains references to deleted commands, and this ticket should remove the reference to mktap from the first line of that file. I will merge #3078 today, so just merge forward, and then you can do it.

Zsh ships [our stub file](http://zsh.git.sourceforge.net/git/gitweb.cgi?p=zsh/zsh;a=blob;f=Completion/Unix/Command/_twisted;h=b3b2427290cc56c37d7343cd134fd24b80e41c8c;hb=HEAD), and we need to maintain compatibility with that for 5 or so years. It's old and crummy and it will break completions horribly for all Twisted commands if anything under twisted/python/zsh is removed. As soon as #3078 is merged I'll ship off the new stub file to the zsh folks, but again, we can't rely on that right away.. people may be using Twisted on systems with old zsh for a long time to come.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [32922]) Branching to 'remove-mktap-5293-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun removed owner|
|-|-|

Alright.  The branch now removes mktap from `twisted-completion.zsh`.  It leaves the other zsh stub dedicated to mktap.  I also deleted the whole .tap section from the application howto.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/remove-mktap-5293-2)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @exarkun|
|-|-|

It's kind of weird to keep tap2deb and tap2rpm. Both otherwise it seems fine, +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [32942]) Merge remove-mktap-5293-2

Author: exarkun
Reviewer: jesstess, therve
Fixes: [#5293](https://github.com/twisted/twisted/issues/5293)

Remove mktap and related support code.
